src/lib/pgen.rs

```diff
@@ -202,3 +203,124 @@ pub fn calc_follows(grm: &Grammar, firsts: &HashMap<String, HashSet<String>>)
         }
     }
 }
+
+#[derive(PartialEq, Eq, Debug)]
+/// Implementation of `LR1Item` which is used to build the state graph of a grammar. Consists of a
+/// left-hand-side `String`, a right-hand-side list of `Symbols` and an integer `dot`, denoting the
+/// current position of the parser inside of `rhs`.
+pub struct LR1Item {
+    pub lhs: String,
+    pub rhs: Vec<Symbol>,
```

Comment:
I might be a dummy, but isn't this continually copying rule symbols? Won't that be really inefficient? Could we not instead number all alternatives from 0..n, and have `lhs` be that integer, from which `rhs` can be inferred?